tarpc/src/trace.rs

```diff
@@ -95,3 +98,46 @@ impl fmt::Display for SpanId {
         Ok(())
     }
 }
+
+#[cfg(feature = "serde1-no-u128")]
+impl serde::Serialize for TraceId {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        use serde::ser::SerializeTupleStruct;
+        let mut tup = serializer.serialize_tuple_struct("TraceId", 1)?;
+        tup.serialize_field(&self.0.to_le_bytes())?;
+        tup.end()
+    }
+}
+
+#[cfg(feature = "serde1-no-u128")]
+impl<'de> serde::Deserialize<'de> for TraceId {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct Visitor;
+
+        impl<'de> serde::de::Visitor<'de> for Visitor {
+            type Value = TraceId;
+
+            fn expecting(&self, formatter: &mut Formatter) -> fmt::Result {
+                formatter.write_str("tuple struct TraceId")
+            }
+
+            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
+            where
+                A: serde::de::SeqAccess<'de>,
+            {
+                let bytes: [u8; 16] = seq.next_element()?.ok_or_else(|| {
+                    serde::de::Error::invalid_length(0, &"tuple struct TraceId with 1 element")
+                })?;
+                Ok(TraceId(u128::from_le_bytes(bytes)))
+            }
+        }
+
+        deserializer.deserialize_tuple_struct("TraceId", 1, Visitor)
+    }
+}
```

Comment:
I think this could be made simpler by using the `#[serde(with)]` attribute on the u128 field. e.g.

```rust
pub struct TraceId(#[cfg_attr(feature = "serde1", serde(with = "u128_serde"))] u128);
```

and then we'd just need a `mod u128_serde` with the serialize and deserialize fns:

```suggestion
#[cfg(feature = "serde1")]
mod u128_serde {
    pub fn serialize<S>(u: &u128, serializer: S) -> Result<S::Ok, S::Error>
        where S: serde::Serializer
    {
        serde::Serialize::serialize(&u.to_le_bytes(), serializer)
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<u128, D::Error>
        where D: serde::Deserializer<'de>
    {
        Ok(u128::from_le_bytes(serde::Deserialize::deserialize(deserializer)?))
    }
}

```

Reply:
that is indeed much simpler, I was too focused on keeping it the same as the expanded Serialize derive. If you don't mind the breaking change I'll update my PR.

Reply:
Ah, that makes sense—thanks for the context! I appreciate you striving to keep the implementation the same.

My preference is to keep things simple. Thanks for the updates!